### PR TITLE
feat: add a question for triggers on scaffold/addCapability

### PR DESCRIPTION
**Pull request by a1exwang**
Add a question to select triggers after ’Select capabilities‘ if user chooses ’Notification‘
image

https://user-images.githubusercontent.com/9698542/158302647-7be65556-9707-4e88-b799-4bb307430545.mp4

https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_workitems/edit/13777227/

### CODE REVIEW

**Comment by a1exwang** [comment]
E2E TEST: n/a

File: packages/fx-core/resource/package.nls.json [hunk 1]
```diff
@@ -99,6 +99,22 @@
   "plugins.frontend.checkSystemTimeTip": "You may get expired credentials, check if your system time is correct.",
   "plugins.frontend.restoreEnvironmentTip": "If you manually updated configuration files (under directory .%s), recover them.",
   "plugins.frontend.baseErrorMessage": "%s Suggestions: %s",
+  "plugins.bot.questionHostTypeTrigger.title": "Choose triggers",
+  "plugins.bot.questionHostTypeTrigger.placeholder": "Choose triggers",
+  "plugins.bot.questionHostTypeTrigger.error.emptySelection": "You have to select at least one trigger.",
+  "plugins.bot.questionHostTypeTrigger.error.hostTypeConflict": "Azure Functions and App Service host types are in conflict",
+  "plugins.bot.triggers.functionsTimer.label": "Timer Trigger",
+  "plugins.bot.triggers.functionsTimer.cliName": "timer",
+  "plugins.bot.triggers.functionsTimer.description": "Timer Trigger",
```
Comment (kimizhu) [inline packages/fx-core/resource/package.nls.json:108]: Give a more meaningful description, you can refer to VS's strings:
https://docs.microsoft.com/en-us/azure/includes/media/functions-vs-tools-create/functions-project-settings.png

Comment (a1exwang) [reply packages/fx-core/resource/package.nls.json]: Updated

**Review by XiaofuHuang — DISMISSED** [review]

**Review by swatDong — DISMISSED** [review]

**Review by jayzhang — DISMISSED** [review]

**Review by swatDong — DISMISSED** [review]

**Review by kimizhu — DISMISSED** [review]

File: packages/fx-core/resource/package.nls.json [hunk 2]
```diff
@@ -113,6 +113,22 @@
   "plugins.frontend.checkSystemTimeTip": "You may get expired credentials, check if your system time is correct.",
   "plugins.frontend.restoreEnvironmentTip": "If you manually updated configuration files (under directory .%s), recover them.",
   "plugins.baseErrorMessage": "%s Suggestions: %s",
+  "plugins.bot.questionHostTypeTrigger.title": "Choose triggers",
```
Comment (IvanJobs) [inline packages/fx-core/resource/package.nls.json:116]: Nit: Better to align wording - it seems multi-env used `Select environments`.

Comment (a1exwang) [reply packages/fx-core/resource/package.nls.json:116]: The wordings are from the PM spec: https://microsoftapc.sharepoint.com/:p:/t/DevDivTeamsDevXProductTeam/EX10MctyIRVFsO6r30kEq94BN1YA8BxLn8mx4kReQy1SbA?e=ZISk6G&nav=eyJzSWQiOjI4NywiY0lkIjozNDE3Mjk1MDE1fQ

File: packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts [hunk 3]
```diff
@@ -48,11 +49,15 @@ export class ScaffoldConfig {
...
-        .filter((item) => item !== undefined) as NotificationTrigger[];
+    const rawHostTypeTriggers = context.answers?.[QuestionNames.BOT_HOST_TYPE_TRIGGER];
+    if (Array.isArray(rawHostTypeTriggers)) {
+      // convert HostTypeTrigger question to trigger name
+      this.triggers = rawHostTypeTriggers
+        .map((hostTypeTrigger) => {
+          const option = HostTypeTriggerOptions.find((option) => option.id === hostTypeTrigger);
+          return option?.trigger;
+        })
+        .filter((item): item is NotificationTrigger => item !== undefined);
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts:60]: Nit: to learn new stuff - what does `item is NotificationTrigger` mean here?

File: packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts [hunk 4]
```diff
@@ -72,22 +77,27 @@ export class ScaffoldConfig {
...
-      // for testing purpose we currently use an environment variable to select hostType.
-      // Change the logic after question model is implemented.
-      if (process.env.TEAMSFX_BOT_HOST_TYPE) {
-        return process.env.TEAMSFX_BOT_HOST_TYPE === "function"
-          ? HostTypes.AZURE_FUNCTIONS
-          : HostTypes.APP_SERVICE;
-      } else {
-        return undefined;
-      }
+      return context.answers
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts:80]: Nit: possible to use something like `return context.answers ?? value`?

File: packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts [hunk 5]
```diff
@@ -72,22 +77,27 @@ export class ScaffoldConfig {
...
+  private static getHostTypeFromHostTypeTriggerQuestion(answers: Inputs): HostType | undefined {
+    // intersection of hostTypeTriggers and HostTypeTriggerOptions
+    const hostTypeTriggers = answers[QuestionNames.BOT_HOST_TYPE_TRIGGER] as unknown;
+    if (Array.isArray(hostTypeTriggers)) {
+      const hostTypes = hostTypeTriggers.map(
+        (item) => HostTypeTriggerOptions.find((option) => option.id === item)?.hostType
+      );
+      return hostTypes ? hostTypes[0] : undefined;
+    } else {
+      return undefined;
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts:97]: Nit: seems similar code logic like above .

File: packages/fx-core/src/plugins/resource/bot/question.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  trigger?: NotificationTrigger;
+}
+
+export const FunctionsTimerTriggerOptionItem: HostTypeTriggerOptionItem = optionWithL10n({
+  id: "functionsTimer",
+  hostType: HostTypes.AZURE_FUNCTIONS,
+  trigger: NotificationTriggers.TIMER,
+});
+
+export const FunctionsHttpTriggerOptionItem: HostTypeTriggerOptionItem = optionWithL10n({
```
Comment (IvanJobs) [inline packages/fx-core/src/plugins/resource/bot/question.ts:24]: Nit: what's the meaning of optionWith`L10n`?

Comment (a1exwang) [reply packages/fx-core/src/plugins/resource/bot/question.ts:24]: I think l10n is a quite common abbreviation of localization similar to i18n for internationalization

File: packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts (hunk 4 above)
Comment (a1exwang) [reply packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts:80]: `context.answers ?? value` is equivalent to `context.answers ? context.answers : value` which is a different logic from the current code.

File: packages/fx-core/resource/package.nls.json [hunk 7]
```diff
@@ -137,6 +137,22 @@
   "plugins.appstudio.manuallySubmitDescription": "Build the Teams app package. You will need to send it to your Teams administrator manually",
   "plugins.appstudio.autoPublishTip": "Install for your organization",
   "plugins.appstudio.autoPublishDescription": "Send your app to your Teams administrator for approval via the Teams App Portal.",
+  "plugins.bot.questionHostTypeTrigger.title": "Choose triggers",
+  "plugins.bot.questionHostTypeTrigger.placeholder": "Choose triggers",
+  "plugins.bot.questionHostTypeTrigger.error.emptySelection": "You have to select at least one trigger.",
+  "plugins.bot.questionHostTypeTrigger.error.hostTypeConflict": "Azure Functions and Azure App Service host types are in conflict.",
+  "plugins.bot.triggers.functionsTimer.label": "Timer Trigger",
+  "plugins.bot.triggers.functionsTimer.cliName": "timer",
```
Comment (IvanJobs) [inline packages/fx-core/resource/package.nls.json:145]: Nit: `cliName` is interesting. Why name it like this, any explanations?

**Review by IvanJobs — APPROVED** [review]
lgtm for the bot part, suppose dev test is passed.

File: packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts (hunk 3 above)
Comment (a1exwang) [reply packages/fx-core/src/plugins/resource/bot/configs/scaffoldConfig.ts:60]: This is called [Type Guard](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates)
```
(item): item is NotificationTrigger => item !== undefined
```
It is to maintain type safety. With `item is NotficationTrigger`, the typescript compiler will know that "If the function returns true, then the return type of this function must be `NotificationTrigger`". So the compiler can automatically deduce the type of the whole expression is `NotificationTrigger[]`:
```
  rawHostTypeTriggers
        .map((hostTypeTrigger) => {
          const option = HostTypeTriggerOptions.find((option) => option.id === hostTypeTrigger);
          return option?.trigger;
        })
        .filter((item): item is NotificationTrigger => item !== undefined);
```

So we don't need explicit type cast (as xxx) which may break the type safety of typescript (i.e. compile time type always equals to runtime type)

File: packages/fx-core/resource/package.nls.json (hunk 7 above)
Comment (a1exwang) [reply packages/fx-core/resource/package.nls.json:145]: All the keys (label, cliName, etc) are directly from `OptionItem` of question model: 
https://github.com/OfficeDev/TeamsFx/blob/829454cccdaba14de331c13830d69c815c1a2454/packages/api/src/types.ts#L23-L48

It is for CLI --argument name

**Review by swatDong — DISMISSED** [review]

**Review by LongOddCode — APPROVED** [review]

**Review by IvanJobs — APPROVED** [review]

**Pull request merged**